File: parser.cpp
/**
 * Copyright (c) 2022 Atsushi Komaba
 *
 * This file is part of Syscraws.
 * Syscraws is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version.
 * Syscraws is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with Syscraws. If not, see <https://www.gnu.org/licenses/>. 
 *
 * @file parser.cpp
 */
#include "parser.hpp"
#include "error.hpp"

static std::unique_ptr<pre_ast::Term> parse_term(lexer::Lexer &);

static std::unique_ptr<pre_ast::Term> parse_factor(lexer::Lexer &lexer){
    std::unique_ptr<pre_ast::Term> ret;
    {
        auto &token_ref = lexer.peek();
        pos::Range pos;
        if(!token_ref){
            ret = nullptr;
        }else if(auto factor = token_ref->factor()){
            ret = std::move(factor);
            pos = std::move(lexer.next()->pos);
        }else if(auto prefix = token_ref->prefix()){
            pos::Range pos_op = std::move(lexer.next()->pos);
            auto operand = parse_factor(lexer);
            if(!operand){
                if(auto token = lexer.next()) throw error::make<error::UnexpectedTokenAfterOperator>(std::move(pos), std::move(token->pos));
                else throw error::make<error::EOFAfterOperator>(std::move(pos));
            }
            pos = pos_op + operand->pos;
            ret = std::make_unique<pre_ast::UnaryOperation>(std::move(pos_op), prefix.value(), std::move(operand));
        }else{
            ret = nullptr;
        }
        if(ret) ret->pos = std::move(pos);
    }
    while(true){
        auto &token_ref = lexer.peek();
        pos::Range pos;
        if(!token_ref) return ret;
        if(auto suffix = token_ref->suffix()){
            pos::Range pos_op = std::move(lexer.next()->pos);
            pos = ret->pos + pos_op;
            ret = std::make_unique<pre_ast::UnaryOperation>(std::move(pos_op), suffix.value(), std::move(ret));
        }else if(auto bracket_type = token_ref->opening_bracket_type()){
            auto pos_open = std::move(lexer.next()->pos);
            bool trailing_comma = true;
            std::vector<std::unique_ptr<pre_ast::Term>> right;
            while(true){
                auto term = parse_term(lexer);
                // ここで term は nullptr の可能性がある
                auto &token = lexer.peek();
                if(token && token->is_comma()){
                    auto pos_comma = std::move(lexer.next()->pos);
                    if(!term) throw error::make<error::EmptyItemInList>(std::move(pos_comma));
                    right.push_back(std::move(term));
                }else{
                    if(term){
                        right.push_back(std::move(term));
                        trailing_comma = false;
                    }
                    break;
                }
            }
            auto close = lexer.next();
            if(!close) throw error::make<error::NoClosingBracket>(std::move(pos_open));
            auto closing_bracket_type = close->closing_bracket_type();
            if(!closing_bracket_type) throw error::make<error::UnexpectedTokenInBracket>(std::move(pos_open), std::move(close->pos));
            if(closing_bracket_type != bracket_type.value()) throw error::make<error::DifferentClosingBracket>(std::move(pos_open), std::move(close->pos));
            pos = (ret ? ret->pos : pos_open) + close->pos;
            ret = std::make_unique<pre_ast::Bracket>(bracket_type.value(), std::move(ret), std::move(right), trailing_comma);
        }else{
            return ret;
        }
        ret->pos = std::move(pos);
    }
}

enum Precedence {
    AssignPrecedence,
    TypePrecedence,
    LogicalOrPrecedence,
    LogicalAndPrecedence,
    ComparisonPrecedence,
    BitOrPrecedence,
    BitXorPrecedence,
    BitAndPrecedence,
    BitShiftPrecedence,
    AddSubPrecedence,
    MulDivRemPrecedence,
    TimeShiftPrecedence,
    MaxPrecedence,
};
static Precedence precedence(pre_ast::BinaryOperator op){
    switch(op){
        case pre_ast::BinaryOperator::ForwardShift:
        case pre_ast::BinaryOperator::BackwardShift:
            return TimeShiftPrecedence;
        case pre_ast::BinaryOperator::Mul:
        case pre_ast::BinaryOperator::Div:
        case pre_ast::BinaryOperator::Rem:
            return MulDivRemPrecedence;
        case pre_ast::BinaryOperator::Add:
        case pre_ast::BinaryOperator::Sub:
            return AddSubPrecedence;
        case pre_ast::BinaryOperator::RightShift:
        case pre_ast::BinaryOperator::LeftShift:
            return BitShiftPrecedence;
        case pre_ast::BinaryOperator::BitAnd:
            return BitAndPrecedence;
        case pre_ast::BinaryOperator::BitXor:
            return BitXorPrecedence;
        case pre_ast::BinaryOperator::BitOr:
            return BitOrPrecedence;
        case pre_ast::BinaryOperator::Equal:
        case pre_ast::BinaryOperator::NotEqual:
        case pre_ast::BinaryOperator::Less:
        case pre_ast::BinaryOperator::LessEqual:
        case pre_ast::BinaryOperator::Greater:
        case pre_ast::BinaryOperator::GreaterEqual:
            return ComparisonPrecedence;
        case pre_ast::BinaryOperator::LogicalAnd:
            return LogicalAndPrecedence;
        case pre_ast::BinaryOperator::LogicalOr:
            return LogicalOrPrecedence;
        case pre_ast::BinaryOperator::Type:
            return TypePrecedence;
        case pre_ast::BinaryOperator::Assign:
        case pre_ast::BinaryOperator::AddAssign:
        case pre_ast::BinaryOperator::SubAssign:
        case pre_ast::BinaryOperator::MulAssign:
        case pre_ast::BinaryOperator::DivAssign:
        case pre_ast::BinaryOperator::RemAssign:
        case pre_ast::BinaryOperator::BitAndAssign:
        case pre_ast::BinaryOperator::BitOrAssign:
        case pre_ast::BinaryOperator::BitXorAssign:
        case pre_ast::BinaryOperator::LeftShiftAssign:
        case pre_ast::BinaryOperator::RightShiftAssign:
        case pre_ast::BinaryOperator::ForwardShiftAssign:
        case pre_ast::BinaryOperator::BackwardShiftAssign:
            return AssignPrecedence;
    }
}
enum class Associativity{
    LeftToRight,
    RightToLeft,
};
static Associativity associativity(int precedence){
    if(precedence == AssignPrecedence) return Associativity::RightToLeft;
    else return Associativity::LeftToRight;
}
static std::unique_ptr<pre_ast::Term> parse_binary_operator(lexer::Lexer &lexer, int current_precedence){
    if(current_precedence == MaxPrecedence){
        return parse_factor(lexer);
    }
    auto left = parse_binary_operator(lexer, current_precedence + 1);
    if(!left) return nullptr;
    bool left_to_right = associativity(current_precedence) == Associativity::LeftToRight;
    while(true){
        auto &op_token = lexer.peek();
        if(!op_token) return left;
        auto op = op_token->infix();
        if(op && precedence(op.value()) == current_precedence){
            auto op_pos = std::move(lexer.next()->pos);
            auto right = parse_binary_operator(lexer, current_precedence + left_to_right);
            pos::Range pos;
            if(right) pos = left->pos + right->pos;
            else if(current_precedence == TypePrecedence) pos = left->pos + op_pos;
            else if(auto token = lexer.next()) throw error::make<error::UnexpectedTokenAfterOperator>(std::move(op_pos), std::move(token->pos));
            else throw error::make<error::EOFAfterOperator>(std::move(op_pos));
            left = std::make_unique<pre_ast::BinaryOperation>(std::move(op_pos), op.value(), std::move(left), std::move(right));
            left->pos = std::move(pos);
            if(left_to_right) continue;
        }
        return left;
    }
}

std::unique_ptr<pre_ast::Term> parse_term(lexer::Lexer &lexer){
    return parse_binary_operator(lexer, 0);
}

template <class EOFError, class UnexpectedTokenError>
std::unique_ptr<token::Token> consume(lexer::Lexer &lexer, bool (token::Token::*cond)() const, pos::Range &arg){
    auto token = lexer.next();
    if(!token) throw error::make<EOFError>(std::move(arg));
    if(!((*token).*cond)()) throw error::make<UnexpectedTokenError>(std::move(arg), std::move(token->pos));
    return token;
}

static std::unique_ptr<pre_ast::Stmt> parse_stmt(lexer::Lexer &lexer){
    auto &token_ref = lexer.peek();
    if(!token_ref){
        // 入力の終端に達した．
        return nullptr;
    }
    if(auto keyword = token_ref->keyword()){
        if(keyword.value() == token::Keyword::If){
            auto pos_if = std::move(lexer.next()->pos);
            auto cond_open = consume<error::UnexpectedEOFAfterKeyword, error::UnexpectedTokenAfterKeyword>(lexer, &token::Token::is_opening_parenthesis, pos_if);
            auto cond = parse_term(lexer);
            if(!cond){
                if(auto token = lexer.next()) throw error::make<error::UnexpectedTokenInBracket>(std::move(cond_open->pos), std::move(token->pos));
                else throw error::make<error::NoClosingBracket>(std::move(cond_open->pos));
            }
            auto cond_close = consume<error::NoClosingBracket, error::UnexpectedTokenInBracket>(lexer, &token::Token::is_closing_parenthesis, cond_open->pos);
            auto stmt_true = parse_stmt(lexer);
            if(!stmt_true){
                auto pos = pos_if + cond_close->pos;
                if(auto token = lexer.next()) throw error::make<error::UnexpectedTokenInControl>(std::move(pos), std::move(token->pos));
                else throw error::make<error::UnexpectedEOFInControl>(std::move(pos));
            }
            auto pos = pos_if + stmt_true->pos;
            std::unique_ptr<pre_ast::Stmt> stmt_false;
            auto &maybe_else = lexer.peek();
            if(maybe_else && maybe_else->keyword() == token::Keyword::Else){
                auto pos_else = std::move(lexer.next()->pos);
                stmt_false = parse_stmt(lexer);
                if(!stmt_false){
                    auto pos_control = pos_if + pos_else;
                    if(auto token = lexer.next()) throw error::make<error::UnexpectedTokenInControl>(std::move(pos_control), std::move(token->pos));
                    else throw error::make<error::UnexpectedEOFInControl>(std::move(pos_control));
                }
                pos += stmt_false->pos;
            }
            auto ret = std::make_unique<pre_ast::If>(std::move(cond), std::move(stmt_true), std::move(stmt_false));
            ret->pos = std::move(pos);
            return ret;
        }else if(keyword.value() == token::Keyword::While){
            auto pos_while = std::move(lexer.next()->pos);
            auto cond_open = consume<error::UnexpectedEOFAfterKeyword, error::UnexpectedTokenAfterKeyword>(lexer, &token::Token::is_opening_parenthesis, pos_while);
            auto cond = parse_term(lexer);
            if(!cond){
                if(auto token = lexer.next()) throw error::make<error::UnexpectedTokenInBracket>(std::move(cond_open->pos), std::move(token->pos));
                else throw error::make<error::NoClosingBracket>(std::move(cond_open->pos));
            }
            auto cond_close = consume<error::NoClosingBracket, error::UnexpectedTokenInBracket>(lexer, &token::Token::is_closing_parenthesis, cond_open->pos);
            auto stmt = parse_stmt(lexer);
            if(!stmt){
                auto pos = pos_while + cond_close->pos;
                if(auto token = lexer.next()) throw error::make<error::UnexpectedTokenInControl>(std::move(pos), std::move(token->pos));
                else throw error::make<error::UnexpectedEOFInControl>(std::move(pos));
            }
            auto pos = pos_while + stmt->pos;
            auto ret = std::make_unique<pre_ast::While>(std::move(cond), std::move(stmt));
            ret->pos = std::move(pos);
            return ret;
        }else if(keyword.value() == token::Keyword::Break){
            auto pos_break = std::move(lexer.next()->pos);
            auto semicolon = consume<error::UnexpectedEOFAfterKeyword, error::UnexpectedTokenAfterKeyword>(lexer, &token::Token::is_semicolon, pos_break);
            auto ret = std::make_unique<pre_ast::Break>();
            ret->pos = pos_break + semicolon->pos;
            return ret;
        }else if(keyword.value() == token::Keyword::Continue){
            auto pos_continue = std::move(lexer.next()->pos);
            auto semicolon = consume<error::UnexpectedEOFAfterKeyword, error::UnexpectedTokenAfterKeyword>(lexer, &token::Token::is_semicolon, pos_continue);
            auto ret = std::make_unique<pre_ast::Continue>();
            ret->pos = pos_continue + semicolon->pos;
            return ret;
        }else if(keyword.value() == token::Keyword::Return){
            auto pos = std::move(lexer.next()->pos);
            auto term = parse_term(lexer);
            if(term) pos += term->pos;
            auto semicolon = consume<error::EOFAtEndOfStmt, error::UnexpectedTokenAtEndOfStmt>(lexer, &token::Token::is_semicolon, pos);
            auto ret = std::make_unique<pre_ast::Return>(std::move(term));
            ret->pos = pos + semicolon->pos;
            return ret;
        }else{
            throw error::make<error::UnexpectedTokenAtBeginningOfStmt>(std::move(lexer.next()->pos));
        }
    }
    auto term = parse_term(lexer);
    auto &token = lexer.peek();
    if(!token){
        // token_ref は nullptr でないが，token は nullptr．
        // よってここで term は空でない
        throw error::make<error::EOFAtEndOfStmt>(std::move(term->pos));
    }else if(token->is_semicolon()){
        auto pos_semicolon = std::move(lexer.next()->pos);
        auto pos = term ? term->pos + pos_semicolon : std::move(pos_semicolon);
        auto ret = std::make_unique<pre_ast::TermStmt>(std::move(term));
        ret->pos = std::move(pos);
        return ret;
    }else if(token->is_opening_brace()){
        auto pos_open = std::move(lexer.next()->pos);
        auto pos = term ? std::move(term->pos) : pos_open.clone();
        std::vector<std::unique_ptr<pre_ast::Stmt>> stmts;
        while(auto stmt = parse_stmt(lexer)) stmts.push_back(std::move(stmt));
        // ここで stmts のどの要素も nullptr でない．
        pos += std::move(consume<error::NoClosingBracket, error::UnexpectedTokenInBracket>(lexer, &token::Token::is_closing_brace, pos_open)->pos);
        auto ret = std::make_unique<pre_ast::Block>(std::move(term), std::move(stmts));
        ret->pos = std::move(pos);
        return ret;
    }else if(!term){
        // parse_stmt() は stmt ::= term? { stmt* } をパースするときに parse_stmt() 自身を呼び出す．
        // このとき，} に達したら nullptr を返す必要がある．
        return nullptr;
    }else{
        throw error::make<error::UnexpectedTokenAtEndOfStmt>(std::move(term->pos), std::move(token->pos));
    }
}

std::unique_ptr<pre_ast::Stmt> parse(lexer::Lexer &lexer){
    if(auto ret = parse_stmt(lexer)){
        lexer.end_stmt();
        return ret;
    }else if(auto token = lexer.next()){
        throw error::make<error::UnexpectedTokenAtBeginningOfStmt>(std::move(token->pos));
    }else{
        return nullptr;
    }
}

File: src/parser.rs
/*
 * Copyright (c) 2023 Atsushi Komaba
 *
 * This file is part of Syscraws.
 * Syscraws is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation, either version 3
 * of the License, or any later version.
 *
 * Syscraws is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Syscraws. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::pre_ast::{Operator, PStmt, PTerm, Stmt, Term};
use std::iter;
mod lexer;
use lexer::Lexer;
mod token;
use token::Token;
mod chars_peekable;
use chars_peekable::CharsPeekable;
mod error;
pub use error::Error;

pub fn parse(input: &str) -> Result<Vec<PStmt>, Error> {
    let mut chars = CharsPeekable::new(input);
    let mut lexer = Lexer::new(&mut chars);
    let mut peeked = lexer.next_token()?;
    iter::from_fn(|| parse_stmt(&mut lexer, &mut peeked).transpose()).collect()
}

fn parse_stmt<'id, 'chars>(
    lexer: &mut Lexer<'id, 'chars>,
    peeked: &mut (usize, Option<Token<'id>>),
) -> Result<Option<PStmt<'id>>, Error> {
    let (first_token_start, Some(ref mut first_token)) = *peeked else {
        return Ok(None);
    };
    match first_token {
        Token::KeywordIf => {
            lexer.consume(peeked)?;
            let if_pos = lexer.range_from(first_token_start);
            let open_start = peeked.0;
            lexer.consume_if_or_err(
                peeked,
                |token| matches!(token, Token::OpeningParenthesis),
                if_pos.clone(),
            )?;
            let open_pos = lexer.range_from(open_start);
            let cond = parse_assign(lexer, peeked)?;
            lexer.consume_if_or_err(
                peeked,
                |token| matches!(token, Token::ClosingParenthesis),
                open_pos,
            )?;
            let stmt_then = parse_stmt(lexer, peeked)?;
            let (else_pos, stmt_else) = if let (else_start, Some(Token::KeywordElse)) = *peeked {
                lexer.consume(peeked)?;
                let else_pos = lexer.range_from(else_start);
                (Some(else_pos), parse_stmt(lexer, peeked)?)
            } else {
                (None, None)
            };
            let pos = lexer.range_from(first_token_start);
            Ok(Some(PStmt::new(
                pos,
                Stmt::If {
                    cond,
                    if_pos,
                    stmt_then: stmt_then.map(Box::new),
                    else_pos,
                    stmt_else: stmt_else.map(Box::new),
                },
            )))
        }
        Token::KeywordWhile => {
            lexer.consume(peeked)?;
            let while_pos = lexer.range_from(first_token_start);
            let open_start = peeked.0;
            lexer.consume_if_or_err(
                peeked,
                |token| matches!(token, Token::OpeningParenthesis),
                while_pos.clone(),
            )?;
            let open_pos = lexer.range_from(open_start);
            let cond = parse_assign(lexer, peeked)?;
            lexer.consume_if_or_err(
                peeked,
                |token| matches!(token, Token::ClosingParenthesis),
                open_pos,
            )?;
            let stmt = parse_stmt(lexer, peeked)?;
            let pos = lexer.range_from(first_token_start);
            Ok(Some(PStmt::new(
                pos,
                Stmt::While {
                    cond,
                    while_pos,
                    stmt: stmt.map(Box::new),
                },
            )))
        }
        Token::KeywordReturn => {
            lexer.consume(peeked)?;
            let term = parse_assign(lexer, peeked)?;
            let pos_excluding_semicolon = lexer.range_from(first_token_start);
            lexer.consume_if_or_err(
                peeked,
                |token| matches!(token, Token::Semicolon),
                pos_excluding_semicolon,
            )?;
            let pos = lexer.range_from(first_token_start);
            Ok(Some(PStmt::new(pos, Stmt::Return(term))))
        }
        _ => {
            let term = parse_assign(lexer, peeked)?;
            match peeked.1 {
                Some(Token::Semicolon) => {
                    lexer.consume(peeked)?;
                    let pos = lexer.range_from(first_token_start);
                    Ok(Some(PStmt::new(pos, Stmt::Term(term))))
                }
                Some(Token::OpeningBrace) => {
                    let open_start = peeked.0;
                    lexer.consume(peeked)?;
                    let open_pos = lexer.range_from(open_start);
                    let mut stmts = Vec::new();
                    loop {
                        if let Some(Token::ClosingBrace) = peeked.1 {
                            lexer.consume(peeked)?;
                            break;
                        } else if let Some(stmt) = parse_stmt(lexer, peeked)? {
                            stmts.push(stmt);
                        } else {
                            return Err(Error::EOFAfter {
                                reason_pos: open_pos,
                            });
                        }
                    }
                    let pos = lexer.range_from(first_token_start);
                    Ok(Some(PStmt::new(
                        pos,
                        Stmt::Block {
                            antecedent: term,
                            stmts,
                        },
                    )))
                }
                Some(_) => {
                    let error_start = peeked.0;
                    lexer.consume(peeked)?;
                    let error_pos = lexer.range_from(error_start);
                    match term {
                        Some(term) => Err(Error::UnexpectedTokenAfter {
                            error_pos,
                            reason_pos: term.pos(),
                        }),
                        None => Err(Error::UnexpectedToken { error_pos }),
                    }
                }
                None => Err(Error::EOFAfter {
                    reason_pos: term.unwrap().pos(),
                }),
            }
        }
    }
}

pub fn parse_assign<'id, 'chars>(
    lexer: &mut Lexer<'id, 'chars>,
    peeked: &mut (usize, Option<Token<'id>>),
) -> Result<Option<PTerm<'id>>, Error> {
    let start = peeked.0;
    let left_hand_side = parse_binary_operator(lexer, peeked)?;
    if let Some(operator) = peeked.1.as_ref().and_then(assignment_operator) {
        let operator_start = peeked.0;
        lexer.consume(peeked)?;
        let operator_pos = lexer.range_from(operator_start);
        let right_hand_side = parse_assign(lexer, peeked)?;
        let pos = lexer.range_from(start);
        Ok(Some(PTerm::new(
            pos,
            Term::Assignment {
                operator,
                operator_pos,
                left_hand_side: left_hand_side.map(Box::new),
                right_hand_side: right_hand_side.map(Box::new),
            },
        )))
    } else {
        Ok(left_hand_side)
    }
}

fn parse_binary_operator<'id, 'chars>(
    lexer: &mut Lexer<'id, 'chars>,
    peeked: &mut (usize, Option<Token<'id>>),
) -> Result<Option<PTerm<'id>>, Error> {
    parse_binary_operator_rec(lexer, peeked, Precedence::first())
}

fn parse_binary_operator_rec<'id, 'chars>(
    lexer: &mut Lexer<'id, 'chars>,
    peeked: &mut (usize, Option<Token<'id>>),
    precedence: Option<Precedence>,
) -> Result<Option<PTerm<'id>>, Error> {
    let Some(precedence) = precedence else {
        return parse_factor(lexer, peeked);
    };
    let start = peeked.0;
    let mut left_operand = parse_binary_operator_rec(lexer, peeked, precedence.next())?;
    while let Some(operator) = peeked
        .1
        .as_ref()
        .and_then(|token| infix_operator(&token, precedence))
    {
        let operator_start = peeked.0;
        lexer.consume(peeked)?;
        let operator_pos = lexer.range_from(operator_start);
        let right_operand = parse_binary_operator_rec(lexer, peeked, precedence.next())?;
        let pos = lexer.range_from(start);
        left_operand = Some(PTerm::new(
            pos,
            Term::BinaryOperation {
                operator_pos,
                left_operand: left_operand.map(Box::new),
                operator,
                right_operand: right_operand.map(Box::new),
            },
        ));
    }
    Ok(left_operand)
}

fn parse_factor<'id, 'chars>(
    lexer: &mut Lexer<'id, 'chars>,
    peeked: &mut (usize, Option<Token<'id>>),
) -> Result<Option<PTerm<'id>>, Error> {
    let (first_token_start, Some(ref mut first_token)) = *peeked else {
        return Ok(None);
    };
    let mut antecedent = 'ant: {
        let term = if let Token::Identifier(name) = *first_token {
            lexer.consume(peeked)?;
            Term::Identifier(name)
        } else if let Token::Number(str) = *first_token {
            lexer.consume(peeked)?;
            let value: String = str.chars().filter(|&ch| ch != '_').collect();
            if value.chars().all(|ch| ch.is_ascii_digit()) {
                Term::Integer(value.parse().unwrap())
            } else {
                Term::Float(value.parse().unwrap())
            }
        } else if let Token::String(components) = first_token {
            let components = std::mem::take(components);
            lexer.consume(peeked)?;
            Term::String(components)
        } else if let Some(operator) = prefix_operator(&first_token) {
            lexer.consume(peeked)?;
            let operator_pos = lexer.range_from(first_token_start);
            let operand = parse_factor(lexer, peeked)?;
            return Ok(Some(PTerm::new(
                lexer.range_from(first_token_start),
                Term::UnaryOperation {
                    operator,
                    operator_pos,
                    operand: operand.map(Box::new),
                },
            )));
        } else {
            break 'ant None;
        };
        Some(PTerm::new(lexer.range_from(first_token_start), term))
    };
    loop {
        let term = {
            let (token_start, Some(ref token)) = *peeked else {
                return Ok(antecedent);
            };
            if let Token::OpeningParenthesis = token {
                lexer.consume(peeked)?;
                let open_pos = lexer.range_from(token_start);
                let mut elements = Vec::new();
                let has_trailing_comma;
                loop {
                    let element = parse_assign(lexer, peeked)?;
                    if let (comma_start, Some(Token::Comma)) = *peeked {
                        lexer.consume(peeked)?;
                        let comma_pos = lexer.range_from(comma_start);
                        elements.push(element.ok_or(comma_pos));
                    } else {
                        if let Some(element) = element {
                            has_trailing_comma = false;
                            elements.push(Ok(element));
                        } else {
                            has_trailing_comma = true;
                        }
                        break;
                    }
                }
                lexer.consume_if_or_err(
                    peeked,
                    |token| matches!(token, Token::ClosingParenthesis),
                    open_pos,
                )?;
                Term::Parenthesized {
                    antecedent: antecedent.map(Box::new),
                    elements,
                    has_trailing_comma,
                }
            } else if let Token::Colon = token {
                lexer.consume(peeked)?;
                let colon_pos = lexer.range_from(token_start);
                let ty = parse_factor(lexer, peeked)?;
                Term::TypeAnnotation {
                    colon_pos,
                    term: antecedent.map(Box::new),
                    ty: ty.map(Box::new),
                }
            } else if let Token::HyphenGreater = token {
                lexer.consume(peeked)?;
                let arrow_pos = lexer.range_from(token_start);
                let ty = parse_factor(lexer, peeked)?;
                Term::ReturnType {
                    arrow_pos,
                    term: antecedent.map(Box::new),
                    ty: ty.map(Box::new),
                }
            } else if let Some(operator) = postfix_operator(token) {
                lexer.consume(peeked)?;
                let operator_pos = lexer.range_from(token_start);
                Term::UnaryOperation {
                    operator,
                    operator_pos,
                    operand: antecedent.map(Box::new),
                }
            } else {
                return Ok(antecedent);
            }
        };
        let pos = lexer.range_from(first_token_start);
        antecedent = Some(PTerm::new(pos, term));
    }
}

fn prefix_operator(token: &Token) -> Option<Operator> {
    match token {
        Token::Plus => Some(Operator::Plus),
        Token::Hyphen => Some(Operator::Minus),
        Token::Slash => Some(Operator::Recip),
        Token::Exclamation => Some(Operator::LogicalNot),
        Token::Tilde => Some(Operator::BitNot),
        Token::DoublePlus => Some(Operator::PreInc),
        Token::DoubleHyphen => Some(Operator::PreDec),
        _ => None,
    }
}
fn postfix_operator(token: &Token) -> Option<Operator> {
    match token {
        Token::DoublePlus => Some(Operator::PostInc),
        Token::DoubleHyphen => Some(Operator::PostDec),
        _ => None,
    }
}

use enum_iterator::Sequence;
#[derive(Clone, Copy, Sequence)]
enum Precedence {
    LogicalOr,
    LogicalAnd,
    Equality,
    Inequality,
    BitOr,
    BitXor,
    BitAnd,
    BitShift,
    AddSub,
    MulDivRem,
    TimeShift,
}

fn infix_operator(token: &Token, precedence: Precedence) -> Option<Operator> {
    match (token, precedence) {
        (Token::TripleGreater, Precedence::TimeShift) => Some(Operator::ForwardShift),
        (Token::TripleLess, Precedence::TimeShift) => Some(Operator::BackwardShift),
        (Token::Asterisk, Precedence::MulDivRem) => Some(Operator::Mul),
        (Token::Slash, Precedence::MulDivRem) => Some(Operator::Div),
        (Token::Percent, Precedence::MulDivRem) => Some(Operator::Rem),
        (Token::Plus, Precedence::AddSub) => Some(Operator::Add),
        (Token::Hyphen, Precedence::AddSub) => Some(Operator::Sub),
        (Token::DoubleGreater, Precedence::BitShift) => Some(Operator::RightShift),
        (Token::DoubleLess, Precedence::BitShift) => Some(Operator::LeftShift),
        (Token::Ampersand, Precedence::BitAnd) => Some(Operator::BitAnd),
        (Token::Circumflex, Precedence::BitXor) => Some(Operator::BitXor),
        (Token::Bar, Precedence::BitOr) => Some(Operator::BitOr),
        (Token::Greater, Precedence::Inequality) => Some(Operator::Greater),
        (Token::GreaterEqual, Precedence::Inequality) => Some(Operator::GreaterEqual),
        (Token::Less, Precedence::Inequality) => Some(Operator::Less),
        (Token::LessEqual, Precedence::Inequality) => Some(Operator::LessEqual),
        (Token::DoubleEqual, Precedence::Equality) => Some(Operator::Equal),
        (Token::ExclamationEqual, Precedence::Equality) => Some(Operator::NotEqual),
        (Token::DoubleAmpersand, Precedence::LogicalAnd) => Some(Operator::LogicalAnd),
        (Token::DoubleBar, Precedence::LogicalOr) => Some(Operator::LogicalOr),
        _ => None,
    }
}

fn assignment_operator(token: &Token) -> Option<Operator> {
    match token {
        Token::Equal => Some(Operator::Assign),
        Token::PlusEqual => Some(Operator::AddAssign),
        Token::HyphenEqual => Some(Operator::SubAssign),
        Token::AsteriskEqual => Some(Operator::MulAssign),
        Token::SlashEqual => Some(Operator::DivAssign),
        Token::PercentEqual => Some(Operator::RemAssign),
        Token::DoubleGreaterEqual => Some(Operator::RightShiftAssign),
        Token::TripleGreaterEqual => Some(Operator::ForwardShiftAssign),
        Token::DoubleLessEqual => Some(Operator::LeftShiftAssign),
        Token::TripleLessEqual => Some(Operator::BackwardShiftAssign),
        Token::AmpersandEqual => Some(Operator::BitAndAssign),
        Token::CircumflexEqual => Some(Operator::BitXorAssign),
        Token::BarEqual => Some(Operator::BitOrAssign),
        _ => None,
    }
}
